### ISSUE TITLE
Question: how to get DNS challenge details again

### ISSUE THREAD

**Issue by centur** [ebekker/ACMESharp#100]
Hi, can anyone give me a hint on how to display DNS challenge again. The one that produced on Complete-ACMEChallenge invocation:

> PS> Complete-ACMEChallenge dns1 -ChallengeType dns-01 -Handler manual
== Manual Challenge Handler - DNS ==
  * Handle Time:      [1/12/2016 1:41:51 PM]
  * Challenge Token:  [xfc0oQahXVqdaBlcZbk5nL8H-GSDFCoQ8LGzOL07qVI]
To complete this Challenge please create a new Resource
Record (RR) with the following characteristics:
  * RR Type:  [TXT]
  * RR Name:  [_acme-challenge.example.com]
  * RR Value: [vNx_fpLgvq0l4rqSATuxhxl9pa155SoeKvNZ98AFB_4]

Also for some reason I have one of my challenge statuses set to invalid when calling 
`(Update-ACMEIdentifier dns1 -ChallengeType dns-01).Challenges`
>...
>Status                 : invalid
>...

Does this means I need to re-submit challenge for that alias or I can fix it somehow (by e.g. setting proper DNS TXT record) ?

**Comment by centur** [comment]
Also - is it allowed to submit challenge before I actually perform the changes - 
e.g. can I call `Complete-ACMEChallenge` then `Submit-ACMEChallenge ` and then, while waiting on some loop polling for `Status:Valid` with `Update-ACMEIdentifier` ,  make my dns changes ? 

Or I must follow the strict order:

`Complete-ACMEChallenge` 
`make DNS changes...`
`Submit-ACMEChallenge `

**Comment by bseddon** [comment]
This command should return details of the DNS challenge.  As you can see, 
(Get-ACMEIdentifier dns1).Challenges Where-Object { $_.Type -eq 'dns-01 }

The answer to your second question is 'yes'.  You can make any modifications you need until you issue the submit command.  It's the submit command that instructs the LE servers to check your DNS txt records for the challenge token.  The complete command is a step to configure ACMESharp with the challenge handler configuration - dns in your case.

**Comment by centur** [comment]
I've read other issues and looks like my initial question make little sense - I failed the challenges for some domains so I need to re-initialize everything from the beginning.
And I probably need to cleanup list of my challenges so invalid ones won't clutter Vault. 

Would be great if we can have a cmdlet for that :)

**Comment by ebekker** [comment]
The issue is, even if you "clean up" some of the local artifacts, the corresponding counterparts found at the LE service are still there (e.g. the Identifier and its related Challenges and their state, any associated Certificates, etc.)

Instead of trying to reset an existing Identifier, you simply create a new one for the same DNS name, and complete its Challenges.  This is how the LE service, and the ACME protocol in general, is designed to operate.

Regarding your initial question, you can re-complete a Challenge that has already been completed by using the `-Regenerate` or `-Repeat` flags depending on what you want to do (see the cmdlet help).

You should only *submit* the Challenge after all changes are completed to satisfy that Challenge.  Once submitted, the ACME server (Let's Encrypt) will test the Challenge -- if it fails, you're done -- that Identifier instance cannot be *resurrected* and you'll need to create a new Identifier and repeat the process.

The `manual` handler in your example spits out details that are needed to make the change, but you must make them manually for your scenario (i.e. for your DNS provider).  There are some handlers that are provided in the ACMESharp library that do the work for you (i.e. AWS R53 for DNS Challenge Type) and they do that work when you invoke the Complete-* cmdlet.

**Comment by ebekker** [comment]
I'm going to close this issue for now, if you have further questions or want to continue the discussion, please reopen.

P.S. Thanks, @bseddon.

**Comment by centur** [comment]
Thanks @ebekker. 
I think my initial mistake was that I've submitted challenges before I actually made these changes so now I have few failed challenges in the vault which are clutter the output. 

Still some cleanup procedure would be handy, because although there is a record about failed validation at LE servers, as an end-user I'm not really interested in my old failed attempts. 

I can raise another issue proposal for that, if you want, something like a cmdlet for retrieving entire history of challenges from LE api and another one to manipulate the local vault list of challenges.

Also as a side question - do I need to submit the challenge every time I need to refresh the certificate or it's just once per domain and then I can re-issue the certificate for that domain as many times as I need ?

**Comment by centur** [comment]
Also I'm not sure I can re-open this issue, at least I can't find how...

**Comment by ebekker** [comment]
My mistake, I didn't realize that GitHub [wouldn't allow](http://stackoverflow.com/a/21333938/5428506) you to reopen if I closed it.

**Comment by ebekker** [comment]
@centur, I've added ticket #102 to add a delete feature to various Vault-stored entities, that I think would address your concerns yet still allow for maintaining historical records by default.  Let me know your thoughts in that ticket.
